{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3e4b0bd8-4465-4c38-ae1f-cf6b713dfaef",
   "metadata": {},
   "outputs": [],
   "source": [
    "def install_packages() -> None:\n",
    "    !pip -q install sadie-antibody\n",
    "    !pip -q install seaborn matplotlib\n",
    "\n",
    "import sys\n",
    "\n",
    "if \"google.colab\" in sys.modules:\n",
    "    install_packages()\n",
    "    get_demo_files()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "88638f13-81c4-4b8d-a418-0214a514a8d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from sadie.renumbering import Renumbering\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "644d47a9-7231-4f36-a430-8c5a7a9b834b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Any, Dict, Optional\n",
    "\n",
    "import pandas as pd\n",
    "from pydantic import BaseModel, Field, root_validator"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc1631ca-1153-4fde-ba00-c2af6f55d4db",
   "metadata": {},
   "source": [
    "## Define the sequences\n",
    "These are the sequences that we are using for this annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a04e995b-6df3-48a5-8b14-0c9e27d3ec84",
   "metadata": {},
   "outputs": [],
   "source": [
    "pg9_seq = \"CAGCGATTAGTGGAGTCTGGGGGAGGCGTGGTCCAGCCTGGGTCGTCCCTGAGACTCTCCTGTGCAGCGTCCGGATTCGACTTCAGTAGACAAGGCATGCACTGGGTCCGCCAGGCTCCAGGCCAGGGGCTGGAGTGGGTGGCATTTATTAAATATGATGGAAGTGAGAAATATCATGCTGACTCCGTATGGGGCCGACTCAGCATCTCCAGAGACAATTCCAAGGATACGCTTTATCTCCAAATGAATAGCCTGAGAGTCGAGGACACGGCTACATATTTTTGTGTGAGAGAGGCTGGTGGGCCCGACTACCGTAATGGGTACAACTATTACGATTTCTATGATGGTTATTATAACTACCACTATATGGACGTCTGGGGCAAAGGGACCACGGTCACCGTCTCGAGC\"\n",
    "pg9_aa = \"QRLVESGGGVVQPGSSLRLSCAASGFDFSRQGMHWVRQAPGQGLEWVAFIKYDGSEKYHADSVWGRLSISRDNSKDTLYLQMNSLRVEDTATYFCVREAGGPDYRNGYNYYDFYDGYYNYHYMDVWGKGTTVTVSS\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49fc8acd-7434-491c-b300-66534f0c64ef",
   "metadata": {},
   "source": [
    "## Renumberring HMM\n",
    "\n",
    "Renumbering module in Sadie uses HMMER to annotate the antibody sequences.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "33ebfa6b-c006-43ed-85e4-ab926d807f1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       PG9                      136 aa                     UNK 01-JAN-1980\n",
      "DEFINITION  .\n",
      "ACCESSION   PG9\n",
      "VERSION     PG9\n",
      "KEYWORDS    .\n",
      "SOURCE      .\n",
      "  ORGANISM  human\n",
      "            .\n",
      "FEATURES             Location/Qualifiers\n",
      "     FWR1            1..24\n",
      "                     /reference=\"human\"\n",
      "                     /gene=\"IGHV3-30*02\"\n",
      "     CDR1            25..32\n",
      "                     /reference=\"human\"\n",
      "                     /gene=\"IGHV3-30*02\"\n",
      "     FWR2            33..49\n",
      "                     /reference=\"human\"\n",
      "                     /gene=\"IGHV3-30*02\"\n",
      "     CDR2            50..57\n",
      "                     /reference=\"human\"\n",
      "                     /gene=\"IGHV3-30*02\"\n",
      "     FWR3            58..95\n",
      "                     /reference=\"human\"\n",
      "                     /gene=\"IGHV3-30*02\"\n",
      "     CDR3            96..125\n",
      "                     /reference=\"human\"\n",
      "                     /gene=\"Missing\"\n",
      "     FWR4            126..136\n",
      "                     /reference=\"human\"\n",
      "                     /gene=\"IGHJ6*04\"\n",
      "ORIGIN\n",
      "        1 qrlvesgggv vqpgsslrls caasgfdfsr qgmhwvrqap gqglewvafi kydgsekyha\n",
      "       61 dsvwgrlsis rdnskdtlyl qmnslrvedt atyfcvreag gpdyrngyny ydfydgyyny\n",
      "      121 hymdvwgkgt tvtvss\n",
      "//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import FeatureLocation, SeqFeature\n",
    "\n",
    "from sadie.renumbering import Renumbering\n",
    "\n",
    "class ProteinSequenceProcessor:\n",
    "    def __init__(self, sequence, seq_name, molecule_type='protein',\n",
    "                 organism=\"Human\", scheme=\"imgt\", region_assign=\"imgt\", \n",
    "                 run_multiproc=False):\n",
    "        self.sequence = sequence\n",
    "        self.scheme = scheme\n",
    "        self.organism = organism\n",
    "        self.name = seq_name\n",
    "        self.type = molecule_type\n",
    "        self.region_assign = region_assign\n",
    "        self.run_multiproc = run_multiproc\n",
    "        self.map_number = {\n",
    "            \"FWR1\": 'v_gene',\n",
    "            \"CDR1\": 'v_gene',\n",
    "            \"FWR2\": 'v_gene',\n",
    "            \"CDR2\": 'v_gene',\n",
    "            \"FWR3\": 'v_gene',\n",
    "            \"CDR3\": 'd_gene',\n",
    "            \"FWR4\": 'j_gene',\n",
    "            \"VGene\": 'v_gene',\n",
    "             \"DGene\": 'd_gene',\n",
    "            \"JGene\": 'j_gene',\n",
    "        }\n",
    "\n",
    "    def process_sequence(self):\n",
    "        renumbering_api = Renumbering(scheme=self.scheme, region_assign=self.region_assign, run_multiproc=self.run_multiproc)\n",
    "\n",
    "        # Run sequence and return renumbering table with sequence_id and sequence\n",
    "        numbering_table = renumbering_api.run_single(self.name, self.sequence)\n",
    "\n",
    "        seq = Seq(self.sequence)\n",
    "        record = SeqRecord(id=self.name, seq=seq, name=self.name)\n",
    "        record.annotations[\"molecule_type\"] = self.type\n",
    "        record.annotations[\"organism\"] = numbering_table.hmm_species[0]\n",
    "\n",
    "        for feature in self.map_number.keys():\n",
    "            _feature = self._get_feature_numbering(numbering_table, feature)\n",
    "            if _feature:\n",
    "                record.features.append(_feature)\n",
    "\n",
    "        return record\n",
    "\n",
    "    def _get_start_stop(self, feature, numbering_table):\n",
    "        start = self.sequence.find(numbering_table[f'{feature.lower()}_aa_no_gaps'][0])\n",
    "        end = start + len(numbering_table[f'{feature.lower()}_aa_no_gaps'][0])\n",
    "        return start, end\n",
    "\n",
    "    def _get_feature_numbering(self, numbering_table, feature):\n",
    "        feature_type = feature\n",
    "        if feature in ['VGene', 'JGene', 'DGene']:\n",
    "            return None  # The numbering scheme is missing these details (will try and get FW and CDR later)\n",
    "        else:\n",
    "            start, end = self._get_start_stop(feature, numbering_table)\n",
    "            qualifier_dict = {'reference': numbering_table.hmm_species[0]}\n",
    "            try:\n",
    "                qualifier_dict['gene'] =  [numbering_table[self.map_number[feature_type]][0]]\n",
    "            except KeyError:\n",
    "                qualifier_dict['gene'] = 'Missing'\n",
    "            location = FeatureLocation(start, end)\n",
    "            _feature = SeqFeature(location, type=feature_type, qualifiers=qualifier_dict)\n",
    "\n",
    "        return _feature\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    seg_name = 'PG9'\n",
    "    pg9_aa = \"QRLVESGGGVVQPGSSLRLSCAASGFDFSRQGMHWVRQAPGQGLEWVAFIKYDGSEKYHADSVWGRLSISRDNSKDTLYLQMNSLRVEDTATYFCVREAGGPDYRNGYNYYDFYDGYYNYHYMDVWGKGTTVTVSS\"\n",
    "    processor = ProteinSequenceProcessor(pg9_aa, seg_name)\n",
    "    genbank_record = processor.process_sequence()\n",
    "    print(genbank_record.format('genbank'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeec0859-efff-4d93-9d36-63efd7bd8da9",
   "metadata": {},
   "source": [
    "## Now we can write the  record to file and visualize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eb241df8-0462-40bb-af71-e3417046c688",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'output/{seg_name}_hmmer.gb',\"w\") as handle:\n",
    "    SeqIO.write(genbank_record,handle,\"genbank\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ebc775a-53f0-4a08-90c9-acb979f52365",
   "metadata": {},
   "source": [
    "### And Visualize the record using Geneious"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aff8d4bf-7bcb-449e-9bc9-8b8ad5dc08b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython import display\n",
    "display.Image(\"images/HMMER_Sadie_Annotation.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25f94db8-226f-4617-aeb4-f5cbfc3fb699",
   "metadata": {},
   "source": [
    "## BLAST Annotation Using SADIE\n",
    "Uses igblastn here to annotate the sequence provided. This is run through the AIRR API, which calls igblast. It can take a single sequence, fasta, or a directory with several files. The output is an AirrTable, which inherits from pandas dataframe and has the same functionalities, plus afew more. We parse the AirTable to get the features added to genbank annotation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79f7580b-b826-4aed-b131-13390fb25a20",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Alternative processor that can take both sequence and nuc ID for annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0d59c88-41f3-4779-91ed-61f5513457f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from sadie.airr import Airr\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import FeatureLocation, SeqFeature\n",
    "from Bio import Entrez\n",
    "\n",
    "class GenBankProcessor:\n",
    "    def __init__(self, email, gene_id, airr_species=\"human\"):\n",
    "        Entrez.email = email\n",
    "        self.gene_id = gene_id\n",
    "        self.airr_api = Airr(airr_species)\n",
    "    \n",
    "    def fetch_genebank_record(self):\n",
    "        try:\n",
    "            handle = Entrez.efetch(db='nucleotide', id=self.gene_id, rettype='gb')\n",
    "            genebank_record = list(SeqIO.parse(handle, \"genbank\"))[0]\n",
    "            return genebank_record\n",
    "        except Exception as e:\n",
    "            print(f\"Error fetching GenBank record: {e}\")\n",
    "            return None\n",
    "\n",
    "    def get_airr_table(self, gene_symbol, sequence):\n",
    "        return self.airr_api.run_single(gene_symbol, sequence)\n",
    "\n",
    "\n",
    "class GenBankFeatureAdder:\n",
    "    def __init__(self, genebank_record, map_dict):\n",
    "        self.genebank_record = genebank_record\n",
    "        self.map_dict = map_dict\n",
    "\n",
    "    def add_feature(self, feature, airr_table):\n",
    "        feature_type = feature\n",
    "        if feature in ['VGene', 'JGene', 'DGene']:\n",
    "            start = int(airr_table[f'{feature_type.lower()[0]}_sequence_start'][0]) - 1\n",
    "            end = int(airr_table[f'{feature_type.lower()[0]}_sequence_end'][0])\n",
    "            qualifier_dict = {\n",
    "                'gene': [airr_table[self.map_dict[feature_type]][0]],\n",
    "                'species': [airr_table['reference_name'][0]]\n",
    "            }\n",
    "        else:\n",
    "            start = int(airr_table[f'{feature_type.lower()}_start'][0]) - 1\n",
    "            end = int(airr_table[f'{feature_type.lower()}_end'][0])\n",
    "            qualifier_dict = {\n",
    "                'gene': [airr_table[self.map_dict[feature_type]][0]],\n",
    "                'reference': [airr_table['reference_name'][0]]\n",
    "            }\n",
    "\n",
    "        location = FeatureLocation(start, end)\n",
    "        _feature = SeqFeature(location, type=feature_type, qualifiers=qualifier_dict)\n",
    "        self.genebank_record.features.append(_feature)\n",
    "def get_record_from_seq(seq, seq_id, seq_name):\n",
    "    seq = Seq(seq)\n",
    "    record = SeqRecord(id=seq_id, seq=seq, name=seq_name)\n",
    "    record.annotations[\"molecule_type\"] = \"DNA\"\n",
    "    return record\n",
    "\n",
    "#genbank_record = get_record_from_seq(seq, seq_id, seq_name)\n",
    "\n",
    "def main(email=None, gene_id=None, seq=None, seq_id=None, seq_name=None):\n",
    "    map_dict = {\n",
    "        \"FWR1\": 'v_call',\n",
    "        \"FWR2\": 'v_call',\n",
    "        \"FWR3\": 'v_call',\n",
    "        \"FWR4\": 'j_call',\n",
    "        \"CDR1\": 'v_call',\n",
    "        \"CDR2\": 'v_call',\n",
    "        \"CDR3\": 'd_call',\n",
    "        \"VGene\": 'v_call',\n",
    "        \"JGene\": 'j_call',\n",
    "        \"DGene\": 'd_call'\n",
    "    }\n",
    "\n",
    "    if seq is not None:\n",
    "        processor = GenBankProcessor(email, gene_id)\n",
    "        genbank_record = get_record_from_seq(seq, seq_id, seq_name)\n",
    "        #print(genbank_record.seq)\n",
    "    else:\n",
    "        processor = GenBankProcessor(email, gene_id)\n",
    "        genbank_record = processor.fetch_genebank_record()\n",
    "        #print('Also got here')\n",
    "\n",
    "    feature_types = [\n",
    "        \"FWR1\",\n",
    "        \"CDR1\",\n",
    "        \"FWR2\",\n",
    "        \"CDR2\",\n",
    "        \"FWR3\",\n",
    "        \"CDR3\",\n",
    "        \"FWR4\",\n",
    "        'VGene',\n",
    "        'DGene',\n",
    "        'JGene'\n",
    "    ]\n",
    "\n",
    "    for feature in feature_types:\n",
    "        airr_table = processor.get_airr_table(feature, genbank_record.seq)\n",
    "        feature_adder = GenBankFeatureAdder(genbank_record, map_dict)\n",
    "        feature_adder.add_feature(feature, airr_table)\n",
    "    print(genbank_record.format('genbank'))\n",
    "    return genbank_record\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    email = 'example@mail.com'\n",
    "    gene_id = 'GU272045.1'\n",
    "    pg9_seq = \"CAGCGATTAGTGGAGTCTGGGGGAGGCGTGGTCCAGCCTGGGTCGTCCCTGAGACTCTCCTGTGCAGCGTCCGGATTCGACTTCAGTAGACAAGGCATGCACTGGGTCCGCCAGGCTCCAGGCCAGGGGCTGGAGTGGGTGGCATTTATTAAATATGATGGAAGTGAGAAATATCATGCTGACTCCGTATGGGGCCGACTCAGCATCTCCAGAGACAATTCCAAGGATACGCTTTATCTCCAAATGAATAGCCTGAGAGTCGAGGACACGGCTACATATTTTTGTGTGAGAGAGGCTGGTGGGCCCGACTACCGTAATGGGTACAACTATTACGATTTCTATGATGGTTATTATAACTACCACTATATGGACGTCTGGGGCAAAGGGACCACGGTCACCGTCTCGAGC\"\n",
    "    id='PG9-Antibody'\n",
    "    seq_name='PG9'\n",
    "\n",
    "    # Fetch the genbank file and annotate \n",
    "    genebank_record = main(seq=pg9_seq, seq_id=id, seq_name=seq_name)\n",
    "    with open(f'output/{id}_airr.gb',\"w\") as handle:\n",
    "        SeqIO.write(genebank_record,handle,\"genbank\")\n",
    "    # Can use a provided sequence\n",
    "    #main(email=email, gene_id=gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce448818-f8a5-4176-8a16-007569a22825",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython import display\n",
    "display.Image(\"images/AIRR_Annotation_incomplete.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d4c6465-0d11-402d-b322-cc1ccbe67f7c",
   "metadata": {},
   "source": [
    "We can also generate one, which includes the rest of the features by fetching from NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07f6a11a-5a54-4f43-aa3f-4e12f0720346",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    email = 'example@mail.com'\n",
    "    gene_id = 'GU272045.1'\n",
    "    #pg9_seq = \"CAGCGATTAGTGGAGTCTGGGGGAGGCGTGGTCCAGCCTGGGTCGTCCCTGAGACTCTCCTGTGCAGCGTCCGGATTCGACTTCAGTAGACAAGGCATGCACTGGGTCCGCCAGGCTCCAGGCCAGGGGCTGGAGTGGGTGGCATTTATTAAATATGATGGAAGTGAGAAATATCATGCTGACTCCGTATGGGGCCGACTCAGCATCTCCAGAGACAATTCCAAGGATACGCTTTATCTCCAAATGAATAGCCTGAGAGTCGAGGACACGGCTACATATTTTTGTGTGAGAGAGGCTGGTGGGCCCGACTACCGTAATGGGTACAACTATTACGATTTCTATGATGGTTATTATAACTACCACTATATGGACGTCTGGGGCAAAGGGACCACGGTCACCGTCTCGAGC\"\n",
    "    #id='PG9-Antibody'\n",
    "    #seq_name='PG9'\n",
    "\n",
    "    #Fetch the genbank file and annotate \n",
    "    #main(seq=pg9_seq, seq_id=id, seq_name=seq_name)\n",
    "\n",
    "    # Can use a provided sequence\n",
    "    genbank_record = main(email=email, gene_id=gene_id)\n",
    "    with open(f'output/{id}_complete.gb',\"w\") as handle:\n",
    "        SeqIO.write(genbank_record,handle,\"genbank\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a01f4b92-16b4-460c-80be-c8271658f7df",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython import display\n",
    "display.Image(\"images/AIRR_Annotation.png\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
